hoomd/state.py

```diff
@@ -158,20 +159,18 @@ def special_pair_types(self):
     @property
     def box(self):
         b = self._cpp_sys_def.getParticleData().getGlobalBox()
-        L = b.getL()
-        return boxdim(Lx=L.x, Ly=L.y, Lz=L.z,
-                      xy=b.getTiltFactorXY(),
-                      xz=b.getTiltFactorXZ(),
-                      yz=b.getTiltFactorYZ(),
-                      dimensions=self._cpp_sys_def.getNDimensions())
-
-    # Set the system box
-    # \param value The new boundaries (a data.boxdim object)
+        return Box._from_cpp(b)
+
     @box.setter
     def box(self, value):
-        if not isinstance(value, boxdim):
-            raise TypeError('box must be a data.boxdim object')
-        self._cpp_sys_def.getParticleData().setGlobalBox(value._getBoxDim())
+        """Set the system box.
+
+        Args:
+            value (hoomd.box.Box): The new box.
+        """
+        if not isinstance(value, Box):
+            raise TypeError('box must be a hoomd.box.Box object')
+        self._cpp_sys_def.getParticleData().setGlobalBox(copy(value)._cpp_obj)
```

Comment:
The new Python Box class uses Lz==0 to flag boxes as 2D. HOOMD's C++ code currently uses the NDimensions member of SystemDefinition. It may make sense to move this parameter over to BoxDim at some point in the future (but not now due to the massive number of changes to C++ files).

However, I think we should add some error checking logic to `setGlobalBox` to give a useful error when the user sets a Lz==0 box and the system is not 2-dimensional.

Reply:
That makes sense for now. I will add the check and throw the error.

Reply:
Actually, we can change the `SystemDefinitiion`'s dimension, so I think raising a warning makes more sense.

Reply:
Good idea!